File: problems/leetcode/0000/22/22.py
# 22. Generate Parentheses
# https://leetcode.com/problems/generate-parentheses/description/

from collections import deque
from typing import *

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        answer = []

        q = deque() # (string, number of opening, number of closing)
        q.append(("", 0, 0))

        while q:
            s, left, right = q.popleft()

            if len(s) == 2 * n:
                answer.append(s)
                continue

            if left < n:
                q.append((s + "(", left + 1, right))
            if right < left:
                q.append((s + ")", left, right + 1))

        return answer

File: problems/leetcode/0000/22/22.cpp
// 22. Generate Parentheses
// https://leetcode.com/problems/generate-parentheses/

using namespace std;

#include <string>
#include <vector>

class Solution {
public:
  int pairs;
  vector<string> answer;
  void bt(string &prev, int left, int right) {
    if (left + right == 2 * pairs) {
      answer.push_back(prev);
      return;
    }
    if (left < pairs) {
      string left_str = prev;
      left_str.push_back('(');
      bt(left_str, left + 1, right);
    }
    if (right < pairs && left > right) {
      string right_str = prev;
      right_str.push_back(')');
      bt(right_str, left, right + 1);
    }
  }
  vector<string> generateParenthesis(int n) {
    pairs = n;
    string init_str = "";
    bt(init_str, 0, 0);
    return answer;
  }
};
